Feat/give user notification during swtichover TIS21-6455

hihilary opened the pull request:
In my opinion, for backend restriction, we can just invoike the permission from role "HEE User Admin", so users won't be able to create/update/delete any accounts during migration.
And for UserMangement UI, we can redirect to another notification page to tell them what's happening. 

TIS21-6455 Restrict the UserMangement access during migration

JosephKelly commented on src/main/java/uk/nhs/hee/tis/usermanagement/config/WebConfig.java:
```diff
@@ -12,10 +13,9 @@
   @Override
   public void addViewControllers(ViewControllerRegistry registry) {
-    registry.addViewController("/").setViewName("forward:/allUsers");
+    registry.addViewController("/").setViewName("forward:/userNotification");
```
I like this approach & it works for the other controllers that we probably also want to replace: `/createUser`, `/updateUser` & `/deleteUser`.

JosephKelly replied:
I would be happy leaving users able to search though, i.e. don't set the view for `/`.

CaiWillis commented:
Just wondering if with a little bit of naming change we could make this into a more generic "user management is under maintenance" feature, rather than have cognito-swtichover specific naming (e.g. parameter name is "underMaintenance" instead of cognitoSwitchover, more generic notification text etc, maybe even parameterize the list of path pattern strings for the interceptor?)

hihilary commented:
> Just wondering if with a little bit of naming change we could make this into a more generic "user management is under maintenance" feature, rather than have cognito-swtichover specific naming (e.g. parameter name is "underMaintenance" instead of cognitoSwitchover, more generic notification text etc, maybe even parameterize the list of path pattern strings for the interceptor?)

Thanks @CaiWillis for the suggestion. 

I've renamed the interceptor and related parameters to use "under maintenance". 

For the request name list, I've added a new parameter in application.properties file, but as we're moving UserManagement UI to Admins-UI, and I'm not very sure if this feature would be reuse, so I prefer not to set up a new parameter on AWS at the moment.

JosephKelly commented on src/main/java/uk/nhs/hee/tis/usermanagement/config/WebConfig.java:
```diff
@@ -1,17 +1,27 @@
+  @Value("${application.under-maintenance}")
+  private boolean underMaintenance;
+
+  @Value("${application.maintenance-requests}")
+  private String[] maintenanceRequestArray;
```
Nice use of `String[]`!

JosephKelly commented on src/main/resources/templates/userNotification.html:
It looks like you've got "closing" html tags.  I think empty `<br/>` tags would be more commonly understood & compatible with very strict browser/parser settings.

JosephKelly commented on src/main/resources/templates/userNotification.html:
```diff
@@ -0,0 +1,24 @@
+    During the switchover, you are not able to create/update/delete any users,
+    and we will notify you on Teams (TIS Support Channel - General) when the switchover is completed.
+    </br></br>
+    We apologise for any inconvenience this may cause, and thank you for your understanding.
```
I'm happy as is.  If you wanted to gain further thoughts, we could explicitly say e.g.

> "If you require additional support during this period modifying user access, please contact us via. usual support channels."

JosephKelly replied:
> We're doing switchover to a new authentication application.</br>
>     During the switchover, you are not able to create/update/delete any users,
>     and we will notify you on Teams (TIS Support Channel - General) when the switchover is completed.

If you are changing anything, you could also change the specific details with a more generic, `Details and updates about this maintenance will be available on Teams (TIS Support Channel - General).`

hihilary replied:
Thanks. These statements are updated.

JosephKelly reviewed (approved):
I agree with your comment about switching over User Management UIs.  That extends to the comments about the HTML too.

hihilary commented:
Did a force-push as I somehow submitted some of my local settings last time. 😢

JosephKelly commented on src/main/resources/templates/userNotification.html:
```diff
@@ -0,0 +1,24 @@
+  <p>
+    Dear users,
+    <br/><br/><br/>
+    User Management is now under maintenance. We're doing switchover to a new authentication application.<br/>
```
You've left in this bit which is still specific to the current planned maintenance.

JosephKelly replied:
> We're doing switchover to a new authentication application.

hihilary replied:
Ah yes! Removed now.
